From 29a0b7a13841bd37624b36a4d1b81b068840b5a0 Mon Sep 17 00:00:00 2001
From: Thiago Macieira <thiago.macieira@intel.com>
Date: Wed, 14 Nov 2018 22:31:01 -0800
Subject: [PATCH] Also include base libraries for AVX2 and AVX512 builds

glibc supports searching for libraries in subdirs according to the CPU
features. On x86-64, those are "haswell" and "avx512_1", as in:

$ ldd /usr/bin/icecc | grep haswell
        libstdc++.so.6 => /usr/lib64/haswell/avx512_1/libstdc++.so.6
        libc.so.6 => /usr/lib64/haswell/libc.so.6
        libm.so.6 => /usr/lib64/haswell/avx512_1/libm.so.6

With this change, we'll include the base library or libraries in the
environment. In the system above, we'll include /usr/lib64/libm.so.6,
/usr/lib64/haswell/libm.so.6 in addition to the libm listed in ldd's
output.
---
 client/icecc-create-env.in | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/client/icecc-create-env.in b/client/icecc-create-env.in
index ce9c6fe..b862636 100755
--- a/client/icecc-create-env.in
+++ b/client/icecc-create-env.in
@@ -109,6 +109,7 @@ add_file ()
 		#         libc.so.6 => /lib/tls/libc.so.6 (0xb7e81000)
 		#         /lib/ld-linux.so.2 (0xb7fe8000)
 		# covering both situations ( with => and without )
+           local lib
            for lib in $(ldd "$path" | sed -n 's,^[^/]*\(/[^ ]*\).*,\1,p'); do
 	      test -f "$lib" || continue
 	      # Check wether the same library also exists in the parent directory,
@@ -116,6 +117,23 @@ add_file ()
 	      local baselib=$(echo "$lib" | sed 's,\(/[^/]*\)/.*\(/[^/]*\)$,\1\2,')
 	      test -f "$baselib" && lib=$baselib
               add_file "$lib"
+
+              # Add the non-haswell and non-avx512_1 libraries too
+              case "$lib" in
+                 */haswell/*|*/avx512_1/*)
+                    ;;
+                 *)
+                    continue
+                    ;;
+              esac
+              local lib_non_avx512=$(echo "$lib" | sed s,/avx512_1/,/,)
+              local lib_non_hsw=$(echo "$lib_non_avx512" | sed s,/haswell/,/,)
+              if [ "$lib" != "$lib_non_avx512" ] && [ -f "$lib_non_avx512" ]; then
+                add_file "$lib_non_avx512"
+              fi
+              if [ "$lib" != "$lib_non_hsw" ] && [ -f "$lib_non_hsw" ]; then
+                add_file "$lib_non_hsw"
+              fi
            done
         fi
     elif test "$is_darwin" = 1; then
-- 
2.19.1

